### PR TITLE
Implement Exam assessment and password protection

### LINKED ISSUE #1564: Lock attempt assessment (or practical exam) with password
For practical exams, it is plausible that instructors would like to restrict the attempting of an assessment (or practical exam). This can be done by locking the attempting of an assessment with a password, set by the instructor. The student is only able to attempt the assessment when he/she enters the password correctly.

### PR DESCRIPTION
**Pull request by allenwq**
Fix #1564 

Exam assessment supports:
   - password protection 
   - publishing all gradings at the same time ( pending to implement )

In order to attempt the assessment, students will need to input the password:
screen shot 2016-10-12 at 2 43 04 pm

The authentication status is session dependent,  the password will be required again if the user signs out or signs in using a different browser.

### CODE REVIEW

**Comment by weiqingtoh** [comment]
LGTM 👍 

Not sure if we need to support time-based exams - this should be able to support it?

File: app/controllers/course/assessment/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                    when 'create'
+                      create_params[:submission_id]
+                    end
+    @submission ||= @assessment.submissions.find(submission_id) if submission_id.present?
+  end
+
+  def redirect_or_create_submission
+    if @submission
+      url = edit_course_assessment_submission_path(current_course, @assessment, @submission)
+      render json: { success: true, submission_url: url }
```
Comment (weiqingtoh) [inline app/controllers/course/assessment/sessions_controller.rb]: How does this differ from redirecting to the URL itself?

**Comment by allenwq** [comment]
@weiqingtoh Time based exams means ?

**Comment by weiqingtoh** [comment]
Means students can only access for a certain timing (from 2pm to 4pm). Or can we do this with assessment open/close dates?

**Comment by allenwq** [comment]
Yes, that's the close date suppose to do.

**Pull request merged**